linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -6286,9 +6309,13 @@ private[optimizer] object OptimizerCore {
    *  whereas `tree.tpe` is always the lowered `RecordType`.
    */
```

Comment:
This specification of `tpe` feels odd. From the doc of `PreTransform`:

> A `PreTransform` has a `tpe` as precisely refined as if a full `transformExpr()` had been performed.

So this is technically correct: a full transformExpr will roll-back any PreTransRecord tree (with the exception of RT long).

Do you know when this is useful in practice?

Reply:
No, right now I'm not sure. I can check later. For now I've addressed all the other comments.

Reply:
How would you like to proceed with this? It seems this is the last open comment on this PR.

From my second look, it seems like before this PR, we needed the `tpe` field to validate if two `PreTransRecordTree`s correspond to the same class. In hindsight, it is not much less obvious to me in the original code that the optimizer was correct in that regard, we're just moving form `rhsOrigType != lhsOrigType` to `lhsStructure.sameClassAs(rhsStructure)`.

This PR introduces a clear distinction between `RefinedType`s and (transient) allocated structures. I think this is an improvement we should keep; it makes many decisions the optimizer makes more explicit (notably in label processing this becomes clear).

However, this feels like a strong hint, that requiring a `tpe: RefinedType` on an arbitrary pre-transform might not be ideal. So maybe we should try an move it up the hierarchy?

Reply:
I've had a look at usages of `PreTransform#tpe` and IMO it's fairly quickly clear why it is useful. For example, take `IsInstanceOf`:

https://github.com/scala-js/scala-js/blob/f82e59f87fbaa10a83cc9502c94e5cae782d18f4/linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala#L558-L576

`PreTransform#tpe` allows the transformation of `IsInstanceOf` to fold the operation if the type is statically know. We want this to happen, if the `PreTransform` ends up as a record.

Another usage I've found is to accurately type bindings: I suspect without the `tpe`, we'd have to roll-back the binding itself in case it needs var replacement (like this, we only need to roll-back the pretransform).

IMO this is sufficient to move forward with this PR, but PTAL as well.